### ISSUE TITLE
Can't set user: as user: $var that is loaded from vars_files

### ISSUE THREAD

**Issue by iliv** [ansible/ansible#2032]
Hello,

consider the two following examples:

```
[1] Doesn't work
$ cat setup/install-server.yml

---

- hosts: backup_server
  vars_files:
   - ../vars/general.yml
  user: $sysuser


$ cat vars/general.yml

---
sysuser: admin


[2] This works
$ cat setup/install-server.yml

---

- hosts: backup_server
  vars_files:
   - ../vars/general.yml
  vars:
   sysuser: admin
  user: $sysuser


$ cat vars/general.yml

---
sysuser: admin
```

Now, I was asked why on Earth would I want to do that?

Basically, I want to specify a user: just once without having to edit all the playbook files when user: changes to something else. Another situation is when I run a playbook against someone else's environment and they don't happen to have this particular system user or don't want to create it. So, I'd love to edit vars/general.yml (the idea is that it's just one file that stores most variables used in a playbook) just once and move on. I know how to use sed, but other people may not.

This is what I was told

[23:06] <@    mdehaan> | user happens before vars_files because the user is an attribute of the play and is currently the same for all hosts in the play
[23:06] <@    mdehaan> | I am fine with that being much more lazily loaded
[23:06] <@    mdehaan> | it just takes some tweaking
[23:06] <@    mdehaan> | it doesn't need to be definite until right before the connection plugin is called
[23:07] <@    mdehaan> | it sounds like it's a lack of a template call really
[23:07] <@    mdehaan> | and nothing more than that
[23:07] <@    mdehaan> | file a ticket and we can see about fixing it

**Comment by progre55** [comment]
Why not pass in the user as an extra-var? As follows:

``` bash
$ cat myplay.yml
---
- hosts: $hosts
  user: $sysuser
  tasks:
  - name: my fancy task
  ...
```

``` bash
$ ansible-playbook myplay.yml -e "hosts=backup_server sysuser=admin"
```

**Comment by mpdehaan** [comment]
Pretty sure you can.   Let's see the output from 'ansible --version' and a
run of your playbook with -v -v -v

**Comment by iliv** [comment]
@mpdehaan 

```
$ ansible --version
ansible 1.0
$ dpkg -l |grep ansi
ii  ansible                            0.9                          Ansible Application
```

It was built from git checkout with 'make deb'

[1] user: $sysuser, where $sysuser is loaded from vars_files

```
$ ansible-playbook -vvv install-server.yml -k -K
SSH password: 
sudo password: 
 ______________________
< PLAY [backup_server] >
 ----------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||


 _________________
< GATHERING FACTS >
 -----------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||


<bitarena-clone> ESTABLISH CONNECTION FOR USER: $sysuser on PORT 22 TO bitarena-clone
fatal: [bitarena-clone] => {'msg': 'FAILED: Authentication failed.', 'failed': True}
```

[2] user: $sysuser, where $sysuser is set as vars: within install-server.yml

```
$ ansible-playbook -vvv setup/install-server.yml -k -K
SSH password: 
sudo password: 
 ______________________
< PLAY [backup_server] >
 ----------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||


 _________________
< GATHERING FACTS >
 -----------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||


<bitarena-clone> ESTABLISH CONNECTION FOR USER: admin on PORT 22 TO bitarena-clone
<bitarena-clone> EXEC /bin/sh -c 'mkdir -p $HOME/.ansible/tmp/ansible-1360354629.56-83994421916378 && chmod a+rx $HOME/.ansible/tmp/ansible-1360354629.56-83994421916378 && echo $HOME/.ansible/tmp/ansible-1360354629.56-83994421916378'
<bitarena-clone> REMOTE_MODULE setup 
<bitarena-clone> PUT /tmp/tmp9_VEQ5 TO /home/admin/.ansible/tmp/ansible-1360354629.56-83994421916378/setup
<bitarena-clone> EXEC /bin/sh -c 'sudo -k && sudo -S -p "[sudo via ansible, key=vijrlopyjtlmidsxbyjoytszkfnnhzcw] password: " -u root /bin/sh -c '"'"'/usr/bin/python /home/admin/.ansible/tmp/ansible-1360354629.56-83994421916378/setup; rm -rf /home/admin/.ansible/tmp/ansible-1360354629.56-83994421916378/ >/dev/null 2>&1'"'"''
ok: [bitarena-clone]
```

**Comment by iliv** [comment]
@progre55 

You must remember what variable you need to pass as well as their values. It becomes even more challenging if multiple people work with the same playbook. If there are more than just one playbook it becomes even more unwieldy. It would require documentation. I'd rather have people look in vars/general.yml and see all the configuration parameters right there, in one place, without having to open documentation located elsewhere and crafting CLI command that clearly complicates matters. A simple ansible-playbook setup/install-server.yml -k -K is elegant and easy on both your eyes and head. Besides, my impression is that traditionally it is meant for temporary changes. What I'm looking for is an option to have as many variables stored in one place as possible (within reason, of course!) so that changing multiple playbook parameters can be done in just one file, or as minimum files as is possible.

Consider, what happens if you run a playbook against someone else's environment? You may want to change username, filesystem paths, etc. Doing that on CLI feels like a wrong place. On the other hand, having stored all that information in vars file, is also self-documenting. So, you do less work, it's easy to find all configurable options in one place, and it's all documented in the very same file.

I like this approach better for what I'm doing right now.

Granted, there are legitimate use cases for extra vars. I think it's a nice feature, I just need something else.

After all, vars: works, why wouldn't vars_files work?

**Comment by mpdehaan** [comment]
So I don't understand the commentary about what is unwieldy or temporary
and such, so let's just keep this conversation on point so it is easy to
read.

Defaults go in "group_vars/all" best.

If you want to guarantee an extra variable is set, "vars_prompt" is great.

If the question is whether you can set the user to SSH from by setting a
magic variable called "sysuser", well, do you have a line that says "user:
$sysuser" ?

You're going to need to show me the playbook itself and not just the output.

gist.github.com may be good for this.

**Comment by mpdehaan** [comment]
Ok, so this is a question about vars_files not being templated into user
land.  Ok, thanks, easy enough.

This should just be us needing to template that variable prior to using it
in the connection plugins.

Thanks for reminding me about the IRC conversation!

**Comment by mpdehaan** [comment]
Fixed in 828119d823720a1d2360d4e52406b669a6f94b49

**Comment by iliv** [comment]
Thanks! It works.
